### PR TITLE
Board Funcs

### PR DESCRIPTION
**Pull request by wastella**
Seperate PR!

### CODE REVIEW

File: tic_tac_toe/src/board.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
+# Board Representation
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: What's this and, more importantly, why is it not a description of the Board class?

File: tic_tac_toe/src/board.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+#
+# -1 = X
+# 0 = No Value
+# 1 = O
+#
+
+
+class Board:
+    def __init__(self, dim):
+        self.values = [None, None, None, None, None, None, None, None, None]
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why does this presume a dim of 3?

File: tic_tac_toe/src/board.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+        self.template = "{}|{}|{}\n-------\n{}|{}|{}\n-------\n{}|{}|{}"
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why does this presume a dim of 3?

File: tic_tac_toe/src/board.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+        self.dim = dim
+
+        self.winner = None
+
+    def convert(self, x, y):
+        """
+        converts a 2-dimensional coordinate to a 1-dimensional coordinate
```
Comment (cestella) [inline tic_tac_toe/src/board.py:17]: `x` and `y` doesn't give me much here.  What does the grid layout look like? Where is `(0,0)` and where is `(2,2)` on a 3x3 grid?

File: tic_tac_toe/src/board.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        Parameters
+        ----------
+        x : integer or float
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why would you support floats?

File: tic_tac_toe/src/board.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+        y : integer or float
+        the first of the two 2-dim coordinates to be operated on
+
+        Returns
+        -------
+        integer or float : the final 1-dim coordinate
+        """
+        if x < 0 or y < 0:
+            return False
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Throw an exception, don't return a bool

File: tic_tac_toe/src/board.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+        # Print out the board, in a way humans can understand
+        return self.template.format(*self.values)
+
+    def validate_move(self, x, y, player):
+        """
+        Takes in a move as 2-dim coordinates, and a player, and validates that it is a valid move
+        Parameters
+        ----------
+        x : integer or float
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Again, why support floats?

File: tic_tac_toe/src/board.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        The first of the two 2-dim coordinates to be operated on
+
+        player : string
+        "X" or "O", which player
+        Returns
+        ---------
+        : boolean
+        Whether or not it is a valid move
+        """
+        _index = self.convert(x, y)
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why is this function not just `return self.values[self.convert(x,y)] == None` ?  If you've already moved into that slot in the grid, it's not a valid move, right?

File: tic_tac_toe/src/board.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+                return True
+
+        elif self.values[_index] == "O":
+            if player == "O":
+                return False
+
+            elif player == "X":
+                return True
+
+    def check_column(self, board, x_or_o):
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: `self` is a `Board` why are you passing in another `board`?

File: tic_tac_toe/src/board.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+        x_or_o : string
+        which player we are checking for
+
+        returns
+        ----------
+        winner : boolean
+        whether or not that player won with a column win
+        """
+        board_width = 3
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why is this not `self.dim`?

File: tic_tac_toe/src/board.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        winner = None
```
Comment (cestella) [inline tic_tac_toe/src/board.py:83]: Think deeper about the default here.  Should it be `None`?

File: tic_tac_toe/src/board.py (hunk 9 above)
Comment (cestella) [inline tic_tac_toe/src/board.py]: This seems like an internal function to the `Board` class and what you really want is a `def check_winner(self)` that calls this function (along with the other 2 you have implemented)

Comment (cestella) [inline tic_tac_toe/src/board.py]: Why doesn't this return the winner rather than take the player as a param?  This should be `def __check_column(self)` and it should return `X`, `O` or `None`

File: tic_tac_toe/src/board.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        winner : boolean
+        Whether or not that player won with a row win
+        """
+        board_width = 3
+        winner = None
+
+        for i in range(0, board_width):
+            values = set()
+            for j in range(0, board_width):
+                index = self.convert(i, j) + 1
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: This is incorrect.

File: tic_tac_toe/src/board.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+                    values.add(board.values[index])
+            if len(values) == 1:
+                if x_or_o == list(values)[0]:
+                    winner = True
+                    break
+                else:
+                    winner = False
+        return winner
+
+    def check_diagonal(self, board, x_or_o):
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Reimplement this with a loop.
Also, this should be `def __check_diagonal(self)` and returns `X`, `O` or `None`

File: tic_tac_toe/src/board_test.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    update_diagonal(board)
+    assert board.check_diagonal(board, "X") == True
+
+def test_validate_move(board):
+    validate_move_update(board)
+    assert board.validate_move(1, 1, "X") == False
+
+board = Board(3)
+
+test_column(board)
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Do not reuse the same board across all tests.  Create the board in each test function.

File: tic_tac_toe/src/board_test.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def test_row(board):
+    update_row(board)
+    assert board.check_row(board, "X") == True
+
+
+def test_diagonal(board):
+    update_diagonal(board)
+    assert board.check_diagonal(board, "X") == True
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Insufficient test cases covered here.  Please cover these (add a new function for each):
* when the diagonal has None in one of the entry and X's and O's in the other
* when the diagonal has None in all of the entries
* when the diagonal has all X's
* when the diagonal has a mix of X's and O's

File: tic_tac_toe/src/board_test.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    board.values[0] = "X"
+
+def test_column(board):
+    update_column(board)
+    assert board.check_column(board, "X") == True
+
+
+def test_row(board):
+    update_row(board)
+    assert board.check_row(board, "X") == True
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Insufficient test cases covered here.  Please cover these (add a new function for each):
* when the row has None in one of the entry and X's and O's in the other
* when the row has None in all of the entries
* when the row has all X's
* when the row has a mix of X's and O's

File: tic_tac_toe/src/board_test.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+def update_row(board):
+    board.values[board.convert(1, 3)] = "X"
+    board.values[board.convert(2, 3)] = "X"
+    board.values[board.convert(3, 3)] = "X"
+
+def validate_move_update(board):
+    board.values[0] = "X"
+
+def test_column(board):
+    update_column(board)
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Insufficient test cases covered here.  Please cover these (add a new function for each):
* when the column has None in one of the entry and X's and O's in the other
* when the column has None in all of the entries
* when the column has all X's
* when the column has a mix of X's and O's

File: tic_tac_toe/src/board_test.py [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    update_row(board)
+    assert board.check_row(board, "X") == True
+
+
+def test_diagonal(board):
+    update_diagonal(board)
+    assert board.check_diagonal(board, "X") == True
+
+def test_validate_move(board):
+    validate_move_update(board)
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Insufficient cases here, Check a valid move and an invalid move.

**Review by cestella — CHANGES_REQUESTED** [review]
Eh, I think I requested some of these changes already, but let's go through this a few times.

File: tic_tac_toe/src/board.py [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+
+class Board:
+    def __init__(self, dim):
+        self.values = [None, None, None, None, None, None, None, None, None]
+
+        self.template = "{}|{}|{}\n-------\n{}|{}|{}\n-------\n{}|{}|{}"
+
+        self.dim = dim
+
+        self.winner = None
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why are we keeping track of the winner as data local to the `Board` class (rather than just having a `is_winner` class that calculates if there's a winner)?

File: tic_tac_toe/src/board.py (hunk 4 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:17]: In the comments, please write out the coordinates of the grid and their positions relative to one another for a 3x3 grid.

File: tic_tac_toe/src/board.py [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+                    values.add(board.values[index])
+            if len(values) == 1:
+                if x_or_o == list(values)[0]:
+                    winner = True
+                    break
+                else:
+                    winner = False
+        return winner
+
+    def check_row(self, board, x_or_o):
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why doesn't this return the winner rather than take the player as a param? This should be `def __check_row(self)` and it should return `X`, `O` or `None`

Comment (cestella) [inline tic_tac_toe/src/board.py]: `self` is a `Board` why are you passing in another board?

Comment (cestella) [inline tic_tac_toe/src/board.py]: This seems like an internal function to the Board class and what you really want is a `def check_winner(self)` that calls this function (along with the other 2 you have implemented)

**Review by cestella — CHANGES_REQUESTED** [review]

**Comment by cestella** [comment]
Also, why is your build failing?

File: tic_tac_toe/src/board.py (hunk 2 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: +1

File: tic_tac_toe/src/board.py (hunk 4 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:17]: On a 3x3 grid, the board would look like:

X|X|X 1
--------
X|X|X 2
--------
X|X|X 3
1 2 3

(1, 1) = 0
(1, 2) = 1
(1, 3) = 2

(2, 1) = 3
(2, 2) = 4
(2, 3) = 5

(3, 1) = 6
(3, 2) = 7
(3, 3) = 8

File: tic_tac_toe/src/board.py (hunk 5 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: fair :+1:

File: tic_tac_toe/src/board.py (hunk 6 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 7 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 8 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 9 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 10 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: :+1:

File: tic_tac_toe/src/board.py (hunk 11 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:83]: Yes, I believe so

File: tic_tac_toe/src/board.py (hunk 9 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

Comment (wastella) [reply tic_tac_toe/src/board.py]: Because its checking winner, not returning winner

File: tic_tac_toe/src/board.py (hunk 12 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: Ok!

File: tic_tac_toe/src/board.py (hunk 20 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: Again, its checking row wins, not figuring out who won

Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

Comment (wastella) [reply tic_tac_toe/src/board.py]: k!

File: tic_tac_toe/src/board.py (hunk 13 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: uhm, I will do the loop thing!

File: tic_tac_toe/src/board.py (hunk 4 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:17]: Ok, if you want to go with 1-index, then you're going to have to change some logic.  I would recommend using 0-index (i.e. `(0,0)`, etc.)

File: tic_tac_toe/src/board.py (hunk 13 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board_test.py (hunk 17 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: :+1:

File: tic_tac_toe/src/board_test.py (hunk 16 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: 👍

File: tic_tac_toe/src/board_test.py (hunk 14 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: Ok!

Comment (wastella) [reply tic_tac_toe/src/board_test.py]: :+1:

File: tic_tac_toe/src/board_test.py (hunk 18 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: 👍

File: tic_tac_toe/src/board_test.py (hunk 15 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: 👍

File: tic_tac_toe/src/board.py (hunk 12 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 20 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 19 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: Becuase we can reuse it through out the class, so it should be a self. var

File: tic_tac_toe/src/board.py (hunk 1 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 4 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:17]: Well, in my convert func i account for the 0, 0 and 1, 1 difference

File: tic_tac_toe/src/board.py [hunk 21]
```diff
@@ -114,22 +114,20 @@ def _check_row(self, x_or_o):
         winner : boolean
         Whether or not that player won with a row win
         """
-        board_width = self.dim
+        board_width = 3
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: you hard coded the board width again!!!

File: tic_tac_toe/src/board.py [hunk 22]
```diff
@@ -151,10 +149,10 @@ def _check_diagonal(self, x_or_o):
         # Check of someone has won
         board_width = self.dim
         winner = None
+        values = set()
 
         for i in range(0, board_width):
-            values = set()
-            index = self.convert(i, i)
+            index = self.convert(i + 1, i + 1)
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: This should be just `i`.  Also, you ignored the other diagonal.  write out the indices and figure out how to compute them.  I'll give you a hint, the indices are:
1. (0, 2)
2. (1, 1)
3. (2, 0)

File: tic_tac_toe/src/board_test.py [hunk 23]
```diff
@@ -172,16 +171,16 @@ def test_validate_move_2(board):
 test_row_4(board)
 
 board = Board(3)
-test_diagonal_1(board)
+# test_diagonal_1(board)
 
 board = Board(3)
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Create hte board object IN the function, please.

File: tic_tac_toe/src/board_test.py [hunk 24]
```diff
@@ -172,16 +171,16 @@ def test_validate_move_2(board):
...
-# test_diagonal_2(board)
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Make sure this checks both diagonals!

File: tic_tac_toe/src/board.py (hunk 21 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 3 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 22 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: :+1:

File: tic_tac_toe/src/board_test.py (hunk 23 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: 👍

File: tic_tac_toe/src/board.py [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+        if x < 0 or y < 0:
+            raise ValueError("Not Proper Inputs")
+
+        elif x > 2 or y > 2:
+            raise ValueError("Not Proper Inputs")
+
+        else:
+            return ((y) * int(self.dim)) + (x)
+
+    def render_board(self):
```
Comment (cestella) [inline tic_tac_toe/src/board.py:36]: This needs to be adaptable to the dimension.  Your template presumes 3x3 board.

**Review by wastella — COMMENTED** [review]
responded

File: tic_tac_toe/src/board.py (hunk 19 above)
Comment (cestella) [reply tic_tac_toe/src/board.py]: who won shouldn't be *part* of the board class.  I totally agree that there should be a function which indicates which player won on the board, but it doesn't need to be a data element of the class.

File: tic_tac_toe/src/board.py (hunk 4 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:17]: Ok, I think we figured this out.  You are going to go for a 0-based approach and translate to 1-based in the calling function.

File: tic_tac_toe/src/board.py (hunk 7 above)
Comment (cestella) [reply tic_tac_toe/src/board.py]: You need to change this docstring still.

File: tic_tac_toe/src/board.py [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+
+        self.values = [None] * dim ** 2
+
+        self.dim = dim
+
+        self.template = ("{}|{}|{}\n-------\n" * int(self.dim - 1)) + "{}|{}|{}"
+
+        self.winner = None
+
+    def convert(self, x, y):
```
Comment (cestella) [inline tic_tac_toe/src/board.py:15]: What if x or y is a float?  Should you cast them to an int?

File: tic_tac_toe/src/board.py (hunk 11 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:83]: So `winner` is true, false or None.  
* How should I interpret `None` in this case?  
* Perhaps in the case when it's None at the end of this function, you throw?  
* Under what conditions would it be None?

File: tic_tac_toe/src/board.py [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+        x_or_o : string
+        Which player we are checking for
+
+        Returns
+        ----------
+        winner : boolean
+        Whether or not that player won with a row win
+        """
+        board_width = self.dim
+        winner = None
```
Comment (cestella) [inline tic_tac_toe/src/board.py:113]: Same comment as in `_check_column`

File: tic_tac_toe/src/board.py [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+        Which player we are checking for
+        Returns
+        ----------
+        winner : boolean
+        Whether or not that player won with a diagonal win
+        """
+
+        # Check of someone has won
+        board_width = self.dim
+        winner = None
```
Comment (cestella) [inline tic_tac_toe/src/board.py:148]: Same comment as in `_check_column`

File: tic_tac_toe/src/board.py [hunk 29]
```diff
@@ -0,0 +1,185 @@
...
+
+        returns
+        ----------
+        winner : boolean
+        whether or not that player won with a column win
+        """
+        board_width = self.dim
+        winner = None
+
+        for i in range(0, board_width):
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Could you name `i` as `row` or `column`?  It would make the code more readable.

File: tic_tac_toe/src/board.py [hunk 30]
```diff
@@ -0,0 +1,185 @@
...
+            values = set()
+            for j in range(0, board_width):
+                index = self.convert(i, j)
+                if self.values[index] != None:
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why not use `validate_move` here?

File: tic_tac_toe/src/board.py [hunk 31]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        Renders the board in a way that the user can easily read it
+        Returns
+        -------
+        string : The board rendered with an actual tic tac toe board template
+        """
+        # Print out the board, in a way humans can understand
+        return self.template.format(*self.values)
+
+    def validate_move(self, x, y, player):
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why do you need `player`?

File: tic_tac_toe/src/board.py [hunk 32]
```diff
@@ -0,0 +1,185 @@
...
+        winner = None
+
+        for i in range(0, board_width):
+            values = set()
+            for j in range(0, board_width):
+                index = self.convert(i, j)
+                if self.values[index] != None:
+                    values.add(self.values[index])
+            if len(values) == 1:
+                if x_or_o == list(values)[0]:
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Rather than converting this set to a list, you could use `next(iter(s))` to grab the first element.

File: tic_tac_toe/src/board_test.py [hunk 33]
```diff
@@ -0,0 +1,229 @@
...
+
+def test_column_4():
+    board = update_column_4()
+    assert (
+        board.check_winner("X", False) == True
+        and board.check_winner("O", False) == True
+    )
+
+
+def test_row_1():
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Add a test with values: `['X','X','X','X','O','X',None,None,None]`

File: tic_tac_toe/src/board.py [hunk 34]
```diff
@@ -0,0 +1,185 @@
...
+                    winner = True
+            else:
+                winner = False
+        return winner
+
+    def _check_row(self, x_or_o):
```
Comment (cestella) [inline tic_tac_toe/src/board.py:96]: This logic is exactly the same as `_check_column` with the exception of one line `index = self.convert(j, i)`.  Can you think of a way to generalize here?  Hint: maybe pass in a function to use to retrieve the index?

File: tic_tac_toe/src/board.py [hunk 35]
```diff
@@ -0,0 +1,185 @@
...
+        winner = None
+        for i in range(0, board_width):
+            values = set()
+            for j in range(0, board_width):
+                index = self.convert(j, i)
+                if self.values[index] != None:
+                    values.add(self.values[index])
+            if len(values) == 1:
+                if x_or_o == list(values)[0]:
+                    winner = True
```
Comment (cestella) [inline tic_tac_toe/src/board.py:122]: Remember to fix the bug here.

File: tic_tac_toe/src/board.py [hunk 36]
```diff
@@ -0,0 +1,185 @@
...
+        ----------
+        winner : boolean
+        Whether or not that player won with a diagonal win
+        """
+
+        # Check of someone has won
+        board_width = self.dim
+        winner = None
+        values = set()
+        values2 = set()
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Better variable name, please.

File: tic_tac_toe/src/board.py [hunk 37]
```diff
@@ -0,0 +1,185 @@
...
+        for i in range(0, board_width):
+            index = self.convert(i, i)
+            index2 = self.convert(i, (-i + 2))
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: better variable name here.  I think you mean to differentiate between the left-to-right diagonal case and the right-to-left diagonal case.

File: tic_tac_toe/src/board.py [hunk 38]
```diff
@@ -0,0 +1,185 @@
...
+            values.add(self.values[index])
+            values2.add(self.values[index2])
+        if len(values) == 1:
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: These two blocks of code are identical with the exception that they're looking at `values` and `values2`.  Can you generalize?  Perhaps with an internal function to `_check_diagonal`?

File: tic_tac_toe/src/board.py [hunk 39]
```diff
@@ -0,0 +1,185 @@
...
+                winner = True
+                return winner
+            else:
+                winner = False
+        else:
+            winner = False
+        return winner
+
+    def check_winner(self, x_or_o, tof):
+        if self._check_column(x_or_o) == tof:
```
Comment (cestella) [inline tic_tac_toe/src/board.py]: Why not just `self._check_column(x_or_o) or self._check_row(x_or_o) or self.check_diagonal(x_or_o)`?

File: tic_tac_toe/src/board_test.py [hunk 40]
```diff
@@ -0,0 +1,229 @@
+from board import Board
+
+
+def update_column_1():
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: These helper function names should be descriptive of what situations they're testing.  I know that you're updating columns and such, but I don't know what situation you're testing.

Also, and more generally, are you reusing any of these functions?  If not, perhaps you should just set these inside of the tests.

File: tic_tac_toe/src/board_test.py [hunk 41]
```diff
@@ -0,0 +1,229 @@
...
+    return board
+
+
+def validate_move_update_1():
+    board = Board(3)
+    board.values[0] = "X"
+    return board
+
+
+def test_column_1():
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Describe the situation that you're testing in a string and name the test function accordingly.  Also, move the implementation of the update function into here.  I shouldn't have to bounce between the top of the file down to here to figure out what situation you're testing.

**Review by cestella — CHANGES_REQUESTED** [review]
Ok, getting better.

File: tic_tac_toe/src/board.py (hunk 19 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 26 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:15]: 👍

File: tic_tac_toe/src/board.py (hunk 31 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: I dont 👍

File: tic_tac_toe/src/board.py (hunk 7 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 29 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 30 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: :+1

File: tic_tac_toe/src/board.py (hunk 32 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 27 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:113]: 👍

File: tic_tac_toe/src/board.py (hunk 28 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:148]: 👍

File: tic_tac_toe/src/board.py (hunk 36 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 37 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: 👍

File: tic_tac_toe/src/board.py (hunk 39 above)
Comment (wastella) [reply tic_tac_toe/src/board.py]: Why?

File: tic_tac_toe/src/board.py (hunk 11 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:83]: 1. Nothing has happened yet, the round is starting
2. Fair Enough
3. None!

File: tic_tac_toe/src/board.py (hunk 25 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:36]: It is, look at self.template

File: tic_tac_toe/src/board.py (hunk 11 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:83]: Actually, because of the way I use `check_winner`, it is alot of the times just `None` because wrong inputs. So im not going to throw at the end if the final product is `None`

File: tic_tac_toe/src/board.py (hunk 34 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:96]: Yes!

File: tic_tac_toe/src/board.py (hunk 35 above)
Comment (wastella) [reply tic_tac_toe/src/board.py:122]: k!

File: tic_tac_toe/src/board_test.py (hunk 33 above)
Comment (wastella) [reply tic_tac_toe/src/board_test.py]: ok!

File: tic_tac_toe/src/board.py (hunk 25 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:36]: Please add a unit test showing the printing of a 4x4 board.

File: tic_tac_toe/src/board.py (hunk 11 above)
Comment (cestella) [reply tic_tac_toe/src/board.py:83]: Ok, fine.  It's weird to mix None's and booleans though.

File: tic_tac_toe/src/board.py (hunk 39 above)
Comment (cestella) [reply tic_tac_toe/src/board.py]: What is `tof` here?  Don't all of these `_check_*` functions return booleans?  Can you add a docstring here?

File: tic_tac_toe/src/board_gen.py [hunk 42]
```diff
@@ -0,0 +1,70 @@
+# Board Functions and Class
+
```
Comment (cestella) [inline tic_tac_toe/src/board_gen.py:2]: What is this class???

File: tic_tac_toe/src/board_gen_test.py [hunk 43]
```diff
@@ -0,0 +1,83 @@
+from board_hyp import Board
```
Comment (cestella) [inline tic_tac_toe/src/board_gen_test.py]: WHat is this test??

File: tic_tac_toe/src/board_test.py [hunk 44]
```diff
@@ -0,0 +1,162 @@
+from board import Board
+
+
+def validate_move_update():
+    board = Board(3)
+    board.values[0] = "X"
+    return board
+
+
+def test_row_1():
```
Comment (cestella) [inline tic_tac_toe/src/board_test.py]: Name these tests something sensible, not numerically.  I have no idea what `test_row_1` is doing from reading the name.  I should know precisely what the test case is from reading the function name.

**Review by cestella — CHANGES_REQUESTED** [review]
One more round.  You seem to have added a new `Board` class?  Should I be reviewing that as well as the old board class?  You also haven't renamed the tests like I asked you.  Also, you have swap files checked in, you should remove the `.swo` files in git.
